pkg/controller/seed-controller-manager/cloud/cloud_controller.go

```diff
@@ -176,46 +179,81 @@ func (r *Reconciler) reconcile(ctx context.Context, log *zap.SugaredLogger, clus
 		}
 	}
 
-	if cluster.Status.CloudMigrationRevision < awsHarcodedAZMigrationRevision {
-		if err := r.migrateAWSMultiAZ(ctx, cluster, prov); err != nil {
-			return nil, err
-		}
-	}
-
-	if _, err := prov.InitializeCloudProvider(cluster, r.updateCluster); err != nil {
+	hanndleProviderError := func(err error) (*reconcile.Result, error) {
 		if kerrors.IsConflict(err) {
 			// In case of conflict we just re-enqueue the item for later
 			// processing without returning an error.
-			r.log.Infow("failed to add finalizer to cluster", "error", err)
+			r.log.Infow("failed to run cloud provider", zap.Error(err))
 			return &reconcile.Result{Requeue: true}, nil
 		}
-		return nil, fmt.Errorf("failed cloud provider init: %v", err)
+
+		return nil, fmt.Errorf("failed cloud provider init: %w", err)
+	}
+
+	// initialize the cloud provider resources that need to exist for the cluster to function;
+	// this will only create the resources once, and becomes a NOP after the first call.
+	//
+	// Some providers support ongoing reconciliations, where missing or broken resources can
+	// automatically be fixed; to prevent excessive API calls to cloud providers, this
+	// reconciliation is only triggered after a certain amount of time has passed since the
+	// last time, configured by the admin.
+	//
+	// To prevent reconciling right after initialization (would cause a bunch of unneeded API
+	// calls), we distinguish early between the providers and use _only_ reconciling when
+	// provider implements it.
+	if betterProvider, ok := prov.(provider.ReconcilingCloudProvider); ok {
+		last := cluster.Status.LastProviderReconciliation
+
+		// default the interval to a safe value
+		interval := datacenter.Spec.ProviderReconciliationInterval
+		if interval == nil || interval.Duration == 0 {
+			interval = &metav1.Duration{Duration: defaults.DefaultCloudProviderReconciliationInterval}
+		}
+
+		// reconcile if the lastTime isn't set (= first time init or a forced reconciliation) or too long ago
+		if last == nil || (interval.Duration > 0 && time.Since(last.Time) >= interval.Duration) {
+			log.Info("Reconciling cloud provider for cluster")
+
+			cluster, err = betterProvider.ReconcileCluster(cluster, r.updateCluster)
+			if err != nil {
+				return hanndleProviderError(err)
+			}
+
+			cluster, err = r.updateCluster(cluster.Name, func(c *kubermaticv1.Cluster) {
+				now := v1.Now()
+				c.Status.LastProviderReconciliation = &now
+			})
+			if err != nil {
+				return nil, fmt.Errorf("failed to set last reconcile timestamp: %w", err)
+			}
+		}
```

Comment:
I think using a RequeueAfter would be sufficient :thinking:

Reply:
Fine by me, I think having the last time a cloud operation failed would be neat, but it's not necessary. A RequeueAfter in `handleProviderError` for all errors would probably help with not hitting APIs too much in error cases.

Reply:
Thinking further about this, maybe you're right. I was wondering if we maybe should the `LastProviderReconciliation` _always_, even if an error occured. This would however dramatically slow down any fixes that could be made. However, consider the errors that can happen during reconciliation: It's pretty much all external stuff, and what are the chances that creating a security group works in 10s if it didn't work this time?

So waiting a bit/much longer in cases of errors could make sense. @kron4eg WDYT?